core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -194,21 +214,32 @@ class GroupCoordinator(val brokerId: Int,
       } else if (!group.supportsProtocols(protocolType, MemberMetadata.plainProtocolSet(protocols))) {
         responseCallback(joinError(memberId, Errors.INCONSISTENT_GROUP_PROTOCOL))
       } else if (group.isPendingMember(memberId)) {
-        // A rejoining pending member will be accepted.
-        addMemberAndRebalance(rebalanceTimeoutMs, sessionTimeoutMs, memberId, clientId, clientHost, protocolType,
-          protocols, group, responseCallback)
+        // A rejoining pending member will be accepted. Note that pending member will never be a static member.
+        assert(groupInstanceId == JoinGroupRequest.EMPTY_GROUP_INSTANCE_ID)
```

Comment:
We generally frown on assertions. It is usually better to raise an exception with a clear message.

Reply:
Could you share more details? The reason for using assertion is to avoid creating invalid state from the code change stage. For example, we have
```private def onGroupLoaded(group: GroupMetadata) {
    group.inLock {
      info(s"Loading group metadata for ${group.groupId} with generation ${group.generationId}")
      assert(group.is(Stable) || group.is(Empty))
      if (groupIsOverCapacity(group)) {
        prepareRebalance(group, s"Freshly-loaded group is over capacity ($groupConfig.groupMaxSize). Rebalacing in order to give a chance for consumers to commit offsets")
      }

      group.allMemberMetadata.foreach(completeAndScheduleNextHeartbeatExpiration(group, _))
    }```
and ```
 // trigger the awaiting join group response callback for all the members after rebalancing
          for (member <- group.allMemberMetadata) {
            assert(member.awaitingJoinCallback != null)
            val joinResult = JoinGroupResult(
              members = if (group.isLeader(member.memberId)) {
```
It would be great if you shed light on the trade-offs on these cases, thank you!

Reply:
@hachikuji Could you give me some guidance on this? Thank you!

Reply:
@guozhangwang I see, so we should choose to throw exception for most times?

Reply:
Yeah I'd suggest so.

Reply:
Yes, exactly. We like stack traces and nice error messages! A lot of these impossible states have a way of becoming more possible over time.